serviceaccount/reconciler.go

```diff
@@ -0,0 +1,143 @@
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/giantswarm/workload-identity-operator-gcp/webhook"
+)
+
+const (
+	AnnotationSecretMetadata  = "kubernetes.io/service-account.name" //#nosec G101
+	AnnotationSecretManagedBy = "app.kubernetes.io/managed-by"       //#nosec  G101
+
+	SecretManagedBy = "workload-identity-operator-gcp" //#nosec G101
+
+	SecretNameSuffix = "google-application-credentials" //#nosec G101
+)
+
+type ServiceAccountReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Logger logr.Logger
+}
+
+//+kubebuilder:rbac:groups=core,resources=serviceaccounts,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=core,resources=serviceaccounts/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=core,resources=serviceaccounts/finalizers,verbs=update
+
+func (r *ServiceAccountReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	serviceAccount := &corev1.ServiceAccount{}
+
+	err := r.Client.Get(ctx, req.NamespacedName, serviceAccount)
+	if err != nil {
+		r.Logger.Error(err, "could not get service account", "service account", req.NamespacedName)
+		return reconcile.Result{}, nil
+	}
+
+	gcpServiceAccount, isGCPAnnotated := serviceAccount.Annotations[webhook.AnnotationGCPServiceAccount]
+	workloadIdentityPool, hasWorkloadIdentity := serviceAccount.Annotations[webhook.AnnotationWorkloadIdentityPoolID]
+	identityProvider, hasIdentityProvider := serviceAccount.Annotations[webhook.AnnotationGCPIdentityProvider]
+
+	if !isGCPAnnotated {
+		message := fmt.Sprintf("Skipping ServiceAccount missing %q annotation", webhook.AnnotationGCPServiceAccount)
+		r.Logger.Info(message, "service-account", req.NamespacedName)
+		return reconcile.Result{}, err
+	}
+
+	if !hasWorkloadIdentity {
+		message := fmt.Sprintf("Skipping ServiceAccount missing %q annotation", webhook.AnnotationWorkloadIdentityPoolID)
+		r.Logger.Info(message, "service-account", req.NamespacedName)
+		return reconcile.Result{}, err
+	}
+
+	if !hasIdentityProvider {
+		message := fmt.Sprintf("Skipping ServiceAccount missing %q annotation", webhook.AnnotationGCPIdentityProvider)
+		r.Logger.Info(message, "service-account", req.NamespacedName)
+		return reconcile.Result{}, err
+	}
+
+	secretName := fmt.Sprintf("%s-%s", serviceAccount.Name, SecretNameSuffix)
+	secret := &corev1.Secret{}
+
+	err = r.Client.Get(ctx, types.NamespacedName{
+		Name:      secretName,
+		Namespace: req.Namespace,
+	}, secret)
+
+	if err != nil && !k8serrors.IsNotFound(err) {
+		r.Logger.Error(err, "secret already exists", "service-account", req.NamespacedName)
+		return reconcile.Result{}, err
+	}
+
+	// Secret already exists, no need to create it again
+	if !secret.CreationTimestamp.IsZero() {
+		return reconcile.Result{}, err
+	}
```

Comment:
I suppose we'd want to update the secret if the annotation changes. Otherwise changing the backing GCP Service Account would only be possible with deleting the Kubernetes ServiceAccount